### PR TITLE
Module#7

### PR DESCRIPTION
**Pull request by Hakimbek**
## `deleteMovie`
Create `deleteMovie` query in `apiSlice.ts` file
```javascript
deleteMovie: builder.mutation<void, number>({
      query: (movieId) => ({
        url: `/movies/${movieId}`,
        method: "DELETE",
      }),
      invalidatesTags: (result, error, id) => [{ type: "Movie", id }],
    }),
```
And use `useDeleteMovieMutation` hook inside `DeleteModal.tsx` in order to delete movie from database
```javascript
const [deleteMovie] = useDeleteMovieMutation();

// Some code

<Button btnType={"btn_danger"} onClick={() => deleteMovie(movie.id)}>
      Confirm
</Button>
```

## `addMovie` and `editMovie`
Create `addMovie` and `qditMovie` query in `apiSlice.ts` file
```javascript
addMovie: builder.mutation<MovieType, IFormValues>({
      query: (movie) => ({
        url: "/movies",
        method: "POST",
        body: movie,
      }),
      invalidatesTags: ["Movie"],
    })
```
```javascript
editMovie: builder.mutation<MovieType, MovieType>({
      query: (movie) => ({
        url: `/movies`,
        method: "PUT",
        body: movie,
      }),
      invalidatesTags: (result, error, arg) => [{ type: "Movie", id: arg.id }],
    }),
```
And use `useAddMovieMutation` and  `useEditMovieMutation` hook inside `ModalForm.tsx` in order to add and edit movie
```javascript
 const [addMovie] = useAddMovieMutation();
 const [editMovie] = useEditMovieMutation();

// Some code

// onSubmit method recieve data as argument and check request method (POST or PUT)
const onSubmit = async (data: IFormValues) => {
    if (typeof data.genres === "string") {
      data.genres = [data.genres];
    }

    try {
      if (Object.keys(movie).length === 0) {
        await addMovie(data).unwrap();
        dispatch(
          setInfoText("The movie has been added to database successfully")
        );
      } else {
        const movieObj = { ...movie, ...data };
        await editMovie(movieObj).unwrap();
        dispatch(setInfoText("The movie has been edited successfully"));
      }
      dispatch(setInfoTitle("Congratulations!"));
    } catch (err) {
      dispatch(setInfoTitle("Error!"));
      dispatch(setInfoText("Something went wrong"));
    }
    handleCloseButton();
    dispatch(handleInfoToggle());
  };

// Some code

<Button onClick={handleSubmit(onSubmit)} btnType={"btn_danger"}>
      Submit
</Button>
```

> I don't think that `onSubmit` function logic is good one. This is the best thing I could do. If you know a good solution, please tell me.

### CODE REVIEW

File: src/components/App.tsx [hunk 1]
```diff
@@ -5,15 +5,18 @@ import MovieDetails from "./MovieDetails/MovieDetails";
...
 import styles from "./app.module.css";
 
 function App() {
-  const movieIsSelected = useAppSelector(selectMovie);
+  const movie = useAppSelector(selectMovie);
 
   return (
     <div className={styles.App}>
-      {movieIsSelected ? <MovieDetails /> : <Header />}
+      {Object.keys(movie).length !== 0 ? <MovieDetails /> : <Header />}
```
Comment (MadinaNishanbaeva) [inline src/components/App.tsx]: Not the best comparison
For movie, define type `null | MovieType` and then you will be able to use just 
```javascript
{movie ? <MovieDetails /> : <Header />}
```

File: src/components/Body/MovieSettings/MovieSettings.tsx [hunk 2]
```diff
@@ -1,6 +1,6 @@
 import { useState } from "react";
 
-import DeleteModal from "../../Modal/DeleteModal";
+import DeleteModal from "../../Modal/DeleteModal/DeleteModal";
```
Comment (MadinaNishanbaeva) [inline src/components/Body/MovieSettings/MovieSettings.tsx:3]: I didn't notice this last time, you need to make sure that when you open the menu on the card, `Movie details` does not open
image

Hint: think about event phases

File: src/components/Modal/InfoModal/InfoModal.tsx [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+import modalStyles from "../modal.module.css";
+
+function InfoModal() {
+  const toggle = useAppSelector(selectInfoToggle);
+  const title = useAppSelector(selectInfoTitle);
+  const text = useAppSelector(selectInfoText);
+  const dispatch = useAppDispatch();
+
+  if (toggle) {
+    return ReactDOM.createPortal(
```
Comment (MadinaNishanbaeva) [inline src/components/Modal/InfoModal/InfoModal.tsx]: Your modals are scattered all over the project

You must have one portal for the entire project, and you represent the content through children

File: src/components/MovieDetails/MovieDetailsBody/MovieDetailsBody.tsx [hunk 4]
```diff
@@ -1,13 +1,12 @@
...
 import Poster from "../../common/Poster/Poster";
-import { MovieType } from "../../../types/movie.type";
 import { getRuntime } from "../../../utils/getRuntime";
 
 import styles from "./movieDetailsBody.module.css";
 
 function MovieDetailsBody() {
-  const movie = useAppSelector(selectMovie) as MovieType;
+  const movie = useAppSelector(selectMovie);
   const runtime = getRuntime(movie.runtime);
```
Comment (MadinaNishanbaeva) [inline src/components/MovieDetails/MovieDetailsBody/MovieDetailsBody.tsx]: Use the getRuntime function directly in the markup

File: src/components/common/Modal/ModalForm/ModalForm.tsx [hunk 5]
```diff
@@ -1,71 +1,147 @@
...
-        <textarea
-          className={styles.modal_textarea}
-          placeholder={"More details"}
-        ></textarea>
+      <ModalTextarea register={register} errors={errors} />
+      <div className={styles.modal_footer}>
+        <Button onClick={() => reset()} btnType={"btn_dark"}>
+          Reset
+        </Button>
+        <Button onClick={handleSubmit(onSubmit)} btnType={"btn_danger"}>
```
Comment (MadinaNishanbaeva) [inline src/components/common/Modal/ModalForm/ModalForm.tsx]: `onSubmit` event should be defined as form attribute

File: src/redux/infoModalSlice.ts [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+});
+
+export const selectInfoToggle = (state: RootState) => state.infoModal.toggle;
+export const selectInfoTitle = (state: RootState) => state.infoModal.title;
+export const selectInfoText = (state: RootState) => state.infoModal.text;
+
+export const { handleInfoToggle, setInfoTitle, setInfoText } =
+  infoModalSlice.actions;
+
+export default infoModalSlice.reducer;
```
Comment (MadinaNishanbaeva) [inline src/redux/infoModalSlice.ts]: you don't need this slice, we do not store value like that in redux

File: src/components/MovieDetails/MovieDetailsHeader/MovieDetailsHeader.tsx [hunk 7]
```diff
@@ -13,7 +13,7 @@ function MovieDetailsHeader() {
     <div className={styles.movieDetails_headerWrapper}>
       <Logo />
       <button
-        onClick={() => dispatch(handleMovieClick(false))}
+        onClick={() => dispatch(handleMovieClick({} as MovieType))}
```
Comment (MadinaNishanbaeva) [inline src/components/MovieDetails/MovieDetailsHeader/MovieDetailsHeader.tsx]: The name does not match the purpose, `seMovieDetails` or smth like that would sound better

**Review by MadinaNishanbaeva — APPROVED** [review]

File: src/components/common/Modal/ModalForm/ModalForm.tsx (hunk 5 above)
Comment (Hakimbek) [reply src/components/common/Modal/ModalForm/ModalForm.tsx]: What is the difference between form and button?

File: src/redux/infoModalSlice.ts (hunk 6 above)
Comment (Hakimbek) [reply src/redux/infoModalSlice.ts]: but how can i display `infoModal` module without redux? Don't I need to store something global to know if the request was successful or failed?

Comment (MadinaNishanbaeva) [reply src/redux/infoModalSlice.ts]: There are so many ways to solve this problem either through react or toolkit. You can leave it as is, but for the future, such data is usually not stored in redux

File: src/components/common/Modal/ModalForm/ModalForm.tsx (hunk 5 above)
Comment (MadinaNishanbaeva) [reply src/components/common/Modal/ModalForm/ModalForm.tsx]: Read [this](https://javascript.info/forms-submit), try to click enter with your form, then give me the answer what's the difference?

Comment (Hakimbek) [reply src/components/common/Modal/ModalForm/ModalForm.tsx]: 1. validate the form before sending
2. pressing `Enter` button triggers `submit` event

Is it correct if I add the `type="submit"` attribute to the button than `onSubmit` to the form

**Pull request merged**